redis-test/test_strings.py

```diff
@@ -106,3 +108,33 @@ def test_select():
         raise Exception('Expect that `SELECT 16` does not work')
     except redis.exceptions.ResponseError as ex:
         assert str(ex) == 'invalid DB index'
+
+def test_pipeline():
+    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    s.connect(("127.0.0.1", 6379))
+    s.send('*1\r\n$4\r\nPING\r\n*1\r\n$4\r\nPING\r\n*1\r\n$4\r\nPING\r\n*1\r\n$4\r\nPING\r\n'.encode())
+    assert s.recv(1024).decode() == '+PONG\r\n+PONG\r\n+PONG\r\n+PONG\r\n'
+    s.close()
+
+def test_pipeline_vs_normal():
+    r = connect()
+    time_pipeline_start = time.time()
+    with r.pipeline(transaction=False) as p:
+        p.set(random_string(10), random_string(100))
+        p.set(random_string(10), random_string(100))
+        p.set(random_string(10), random_string(100))
+        p.set(random_string(10), random_string(100))
+        p.set(random_string(10), random_string(100))
+        p.set(random_string(10), random_string(100))
+        p.execute()
+    time_pipeline = time.time() - time_pipeline_start 
+    time_normal_start = time.time()
+    r.set(random_string(10), random_string(100))
+    r.set(random_string(10), random_string(100))
+    r.set(random_string(10), random_string(100))
+    r.set(random_string(10), random_string(100))
+    r.set(random_string(10), random_string(100))
+    r.set(random_string(10), random_string(100))
+    time_normal = time.time() - time_normal_start
+
+    assert time_pipeline < time_normal
```

Comment:
This test can be "flaky", meaning it will randomly fail, just because of timing issues. E.g., the pipeline should have slightly lower latency - but if in parallel something else runs on the test machine, suddenly it can run a bit slower and the test will fail.
So I don't think we can have such a test. 

On the other hand, there are things we **can** test, and you don't test. We can and should test for correctness - namely that if we do 10 writes to different keys, all of them took effect and the response includes all of them (and we don't e.g., miss one of them) and in the right order. We should also test that  operations on the same key (e.g., read and write) are not reordered.

Reply:
Thanks for your suggestion. I'll fix it later.